File: core/src/com/jupiter/europa/screen/dialog/LoadGameDialog.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.jupiter.europa.screen.dialog;

import com.badlogic.gdx.Input;
import com.badlogic.gdx.scenes.scene2d.InputEvent;
import com.badlogic.gdx.scenes.scene2d.ui.*;
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener;
import com.badlogic.gdx.scenes.scene2d.utils.SpriteDrawable;
import com.badlogic.gdx.utils.Align;
import com.jupiter.europa.EuropaGame;
import com.jupiter.europa.scene2d.ui.EuropaButton;
import com.jupiter.europa.scene2d.ui.EuropaButton.ClickEvent;
import com.jupiter.europa.scene2d.ui.ObservableDialog;
import com.jupiter.europa.screen.MainMenuScreen;

import static com.jupiter.europa.screen.MainMenuScreen.*;

/**
 *
 * @author Nathan Templon
 */
public class LoadGameDialog extends ObservableDialog {

    // Enumerations
    public enum LoadGameExitStates {

        LOAD,
        CANCEL
    }


    // Constants
    private static final String DIALOG_NAME = "";


    // Static Methods
    private static Skin getDefaultSkin() {
        return MainMenuScreen.getMainMenuSkin();
    }


    // Fields
    private final Skin skin = getDefaultSkin();

    private Table mainTable;
    private Label loadGameLabel;
    private Table listTable;
    private List gameList;
    private ScrollPane gameListPane;
    private Table buttonTable;
    private EuropaButton okButton;
    private EuropaButton cancelButton;
    private EuropaButton deleteButton;
    
    private float width, height;

    private String gameToLoad = "";
    private LoadGameExitStates exitState;

    private ConfirmDeleteSaveDialog confirmDeleteDialog;


    // Properties
    public final String getGameToLoad() {
        return this.gameToLoad;
    }

    public final LoadGameExitStates getExitState() {
        return this.exitState;
    }


    // Initialization
    public LoadGameDialog() {
        super(DIALOG_NAME, getDefaultSkin());

        this.initComponents();
    }


    // Public Methods
    @Override
    public void setSize(float width, float height) {
        super.setSize(width, height);
        if (this.confirmDeleteDialog != null) {
            this.confirmDeleteDialog.setSize(width, this.height);
        }
        
        this.width = width;
        this.height = height;
    }


    // Private Methods
    private void initComponents() {
        this.mainTable = new Table();
        
        this.listTable = new Table();
        this.loadGameLabel = new Label("Save Games", skin.get(DEFAULT_KEY, Label.LabelStyle.class));
        this.gameList = new List(skin.get(DEFAULT_KEY, List.ListStyle.class));
        this.gameList.setItems((Object[]) EuropaGame.game.getSaveNames());
        this.gameListPane = new ScrollPane(this.gameList, skin.get(DEFAULT_KEY, ScrollPane.ScrollPaneStyle.class));
        
        this.listTable.add(this.gameListPane).pad(MainMenuScreen.LIST_WRAPPER_PADDING).expand().fill();
        this.listTable.background(skin.get(LIST_BACKGROUND_KEY, SpriteDrawable.class));

        this.okButton = new EuropaButton("Accept", skin.get(DEFAULT_KEY, TextButton.TextButtonStyle.class));
        this.okButton.addClickListener(this::onLoadClick);

        this.cancelButton = new EuropaButton("Cancel", skin.get(DEFAULT_KEY, TextButton.TextButtonStyle.class));
        this.cancelButton.addClickListener(this::onCancelClick);

        this.deleteButton = new EuropaButton("Delete", skin.get(DEFAULT_KEY, TextButton.TextButtonStyle.class));
        this.deleteButton.addClickListener(this::onDeleteClick);

        this.buttonTable = new Table();
        this.buttonTable.add(this.cancelButton).space(MainMenuScreen.COMPONENT_SPACING).width(MainMenuScreen.DIALOG_BUTTON_WIDTH).right().expandX();
        this.buttonTable.add(this.deleteButton).space(MainMenuScreen.COMPONENT_SPACING).width(MainMenuScreen.DIALOG_BUTTON_WIDTH).right();
        this.buttonTable.add(this.okButton).space(MainMenuScreen.COMPONENT_SPACING).width(MainMenuScreen.DIALOG_BUTTON_WIDTH).right();
        
        this.mainTable.pad(MainMenuScreen.TABLE_PADDING);

        this.mainTable.add(this.loadGameLabel).center().left();
        this.mainTable.row();
        this.mainTable.add(this.listTable).expandY().fill();
        this.mainTable.row();
        this.mainTable.add(this.buttonTable).space(MainMenuScreen.COMPONENT_SPACING).right().expandX().fillX();
        this.mainTable.row();
        
        this.mainTable.background(this.skin.get(MainMenuScreen.DIALOG_BACKGROUND_KEY, SpriteDrawable.class));

        this.getContentTable().add(this.mainTable).center().expandY().fillY().width(MainMenuScreen.DIALOG_WIDTH);
    }

    private void onLoadClick(ClickEvent event) {
        this.gameToLoad = this.gameList.getSelected().toString();
        this.exitState = LoadGameExitStates.LOAD;
        this.hide();
    }

    private void onCancelClick(ClickEvent event) {
        this.gameToLoad = "";
        this.exitState = LoadGameExitStates.CANCEL;
        this.hide();
    }

    private void onDeleteClick(ClickEvent event) {
        this.confirmDeleteDialog = new ConfirmDeleteSaveDialog(this.gameList.getSelected().toString(), this.skin);
        this.confirmDeleteDialog.addDialogListener(this::onConfirmDeleteSaveDialogClose, DialogEvents.HIDDEN);
        this.confirmDeleteDialog.show(this.getStage());
        this.confirmDeleteDialog.setSize(this.width, this.height);
    }

    private void onConfirmDeleteSaveDialogClose(DialogEventArgs args) {
        if (this.confirmDeleteDialog.getExitState() == ConfirmDeleteSaveDialog.ConfirmDeleteExitStates.DELETE) {
            EuropaGame.game.deleteSave(this.gameList.getSelected().toString());
            this.gameList.setItems((Object[]) EuropaGame.game.getSaveNames());
            this.okButton.setDisabled(EuropaGame.game.getSaveNames().length == 0);
        }
    }


    // Nested Classes
    private static class ConfirmDeleteSaveDialog extends ObservableDialog {

        // Enumerations
        public enum ConfirmDeleteExitStates {

            DELETE,
            CANCEL
        }


        // Constants
        private static final String DIALOG_NAME = "";


        // Fields
        private final String saveName;
        private final Skin skin;

        private Label titleLabel;
        private Table mainTable;
        private TextButton yesButton;
        private TextButton noButton;

        private ConfirmDeleteExitStates exitState;


        // Properties
        public String getSaveToDelete() {
            return this.saveName;
        }

        public ConfirmDeleteExitStates getExitState() {
            return this.exitState;
        }


        // Initialization
        private ConfirmDeleteSaveDialog(String saveName, Skin skin) {
            super(DIALOG_NAME, skin.get(MainMenuScreen.POPUP_DIALOG_STYLE_KEY, WindowStyle.class));

            this.saveName = saveName;
            this.skin = skin;

            this.initComponents();
        }
        
        
        // Public Methods
        @Override
        public void setSize(float width, float height) {
            super.setSize(width, height);
        }


        // Private Methods
        private void initComponents() {
            this.titleLabel = new Label("Are you sure you want to delete the save game \"" + this.saveName + "\"?", skin.get(INFO_STYLE_KEY,
                    Label.LabelStyle.class));
            this.titleLabel.setWrap(true);
            this.titleLabel.setAlignment(Align.center);
            
            this.yesButton = new TextButton("Yes", skin.get(DEFAULT_KEY, TextButton.TextButtonStyle.class));
            this.yesButton.addListener(new ClickListener() {
                @Override
                public void clicked(InputEvent event, float x, float y) {
                    if (event.getButton() == Input.Buttons.LEFT && !ConfirmDeleteSaveDialog.this.yesButton.isDisabled()) {
                        ConfirmDeleteSaveDialog.this.onYesButtonClick();
                    }
                    ConfirmDeleteSaveDialog.this.yesButton.setChecked(false);
                }
            });
            
            this.noButton = new TextButton("No", skin.get(DEFAULT_KEY, TextButton.TextButtonStyle.class));
            this.noButton.addListener(new ClickListener() {
                @Override
                public void clicked(InputEvent event, float x, float y) {
                    if (event.getButton() == Input.Buttons.LEFT && !ConfirmDeleteSaveDialog.this.noButton.isDisabled()) {
                        ConfirmDeleteSaveDialog.this.onNoButtonClick();
                    }
                    ConfirmDeleteSaveDialog.this.noButton.setChecked(false);
                }
            });
            
            this.mainTable = new Table();
            this.mainTable.add(this.titleLabel).colspan(2).width(MainMenuScreen.DIALOG_WIDTH).expandX().fillX();
            this.mainTable.row();
            this.mainTable.add(this.noButton).width(MainMenuScreen.DIALOG_BUTTON_WIDTH).space(MainMenuScreen.COMPONENT_SPACING).right();
            this.mainTable.add(this.yesButton).width(MainMenuScreen.DIALOG_BUTTON_WIDTH).space(MainMenuScreen.COMPONENT_SPACING).left();
            
            this.mainTable.padLeft(MainMenuScreen.TABLE_PADDING);
            this.mainTable.padRight(MainMenuScreen.TABLE_PADDING);
            
            this.mainTable.background(this.skin.get(MainMenuScreen.POPUP_BACKGROUND_KEY, SpriteDrawable.class));
            this.getContentTable().add(this.mainTable).row();
        }

        private void onYesButtonClick() {
            this.exitState = ConfirmDeleteExitStates.DELETE;
            this.hide();
        }

        private void onNoButtonClick() {
            this.exitState = ConfirmDeleteExitStates.CANCEL;
            this.hide();
        }

    }

}

File: core/src/com/jupiter/europa/screen/dialog/LoadGameDialog.kt
/*
 * The MIT License
 *
 * Copyright 2015 Nathan Templon.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package com.jupiter.europa.screen.dialog

import com.badlogic.gdx.Input
import com.badlogic.gdx.scenes.scene2d.InputEvent
import com.badlogic.gdx.scenes.scene2d.ui.*
import com.badlogic.gdx.scenes.scene2d.utils.ClickListener
import com.badlogic.gdx.scenes.scene2d.utils.SpriteDrawable
import com.badlogic.gdx.utils.Align
import com.jupiter.europa.EuropaGame
import com.jupiter.europa.scene2d.ui.EuropaButton
import com.jupiter.europa.scene2d.ui.EuropaButton.ClickEvent
import com.jupiter.europa.scene2d.ui.ObservableDialog
import com.jupiter.europa.screen.MainMenuScreen
import com.badlogic.gdx.scenes.scene2d.ui.List as GuiList
import com.badlogic.gdx.utils.Array as GdxArray

/**

 * @author Nathan Templon
 */
public class LoadGameDialog : ObservableDialog(LoadGameDialog.DIALOG_NAME, LoadGameDialog.getDefaultSkin()) {

    // Enumerations
    public enum class LoadGameExitStates {

        LOAD,
        CANCEL
    }


    // Fields
    private val skinInternal = getDefaultSkin()

    private var mainTable: Table? = null
    private var loadGameLabel: Label? = null
    private var listTable: Table? = null
    private var gameList: GuiList<String>? = null
    private var gameListPane: ScrollPane? = null
    private var buttonTableInternal: Table? = null
    private var okButton: EuropaButton? = null
    private var cancelButton: EuropaButton? = null
    private var deleteButton: EuropaButton? = null

    private var widthInternal: Float = 0.toFloat()
    private var heightInternal: Float = 0.toFloat()

    public var gameToLoad: String = ""
        private set
    public var exitState: LoadGameExitStates? = null
        private set

    private var confirmDeleteDialog: ConfirmDeleteSaveDialog? = null


    init {

        this.initComponents()
    }


    // Public Methods
    override fun setSize(width: Float, height: Float) {
        super.setSize(width, height)
        if (this.confirmDeleteDialog != null) {
            this.confirmDeleteDialog!!.setSize(width, this.heightInternal)
        }

        this.widthInternal = width
        this.heightInternal = height
    }


    // Private Methods
    private fun initComponents() {
        this.mainTable = Table()

        this.listTable = Table()
        this.loadGameLabel = Label("Save Games", skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<Label.LabelStyle>()))
        this.gameList = GuiList(skinInternal.get<GuiList.ListStyle>(MainMenuScreen.DEFAULT_KEY, javaClass<GuiList.ListStyle>()))
        this.gameList?.setItems(GdxArray(EuropaGame.game.getSaveNames()))
        this.gameListPane = ScrollPane(this.gameList, skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<ScrollPane.ScrollPaneStyle>()))

        this.listTable!!.add<ScrollPane>(this.gameListPane).pad(MainMenuScreen.LIST_WRAPPER_PADDING.toFloat()).expand().fill()
        this.listTable!!.background(skinInternal.get(MainMenuScreen.LIST_BACKGROUND_KEY, javaClass<SpriteDrawable>()))

        this.okButton = EuropaButton("Accept", skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<TextButton.TextButtonStyle>()))
        this.okButton!!.addClickListener { args -> this.onLoadClick(args) }

        this.cancelButton = EuropaButton("Cancel", skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<TextButton.TextButtonStyle>()))
        this.cancelButton!!.addClickListener { args -> this.onCancelClick(args) }

        this.deleteButton = EuropaButton("Delete", skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<TextButton.TextButtonStyle>()))
        this.deleteButton!!.addClickListener { args -> this.onDeleteClick(args) }

        this.buttonTableInternal = Table()
        this.buttonTableInternal!!.add<EuropaButton>(this.cancelButton).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).width(MainMenuScreen.DIALOG_BUTTON_WIDTH.toFloat()).right().expandX()
        this.buttonTableInternal!!.add<EuropaButton>(this.deleteButton).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).width(MainMenuScreen.DIALOG_BUTTON_WIDTH.toFloat()).right()
        this.buttonTableInternal!!.add<EuropaButton>(this.okButton).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).width(MainMenuScreen.DIALOG_BUTTON_WIDTH.toFloat()).right()

        this.mainTable!!.pad(MainMenuScreen.TABLE_PADDING.toFloat())

        this.mainTable!!.add<Label>(this.loadGameLabel).center().left()
        this.mainTable!!.row()
        this.mainTable!!.add<Table>(this.listTable).expandY().fill()
        this.mainTable!!.row()
        this.mainTable!!.add<Table>(this.buttonTableInternal).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).right().expandX().fillX()
        this.mainTable!!.row()

        this.mainTable!!.background(this.skinInternal.get(MainMenuScreen.DIALOG_BACKGROUND_KEY, javaClass<SpriteDrawable>()))

        this.getContentTable().add<Table>(this.mainTable).center().expandY().fillY().width(MainMenuScreen.DIALOG_WIDTH.toFloat())
    }

    private fun onLoadClick(event: ClickEvent) {
        this.gameToLoad = this.gameList!!.getSelected().toString()
        this.exitState = LoadGameExitStates.LOAD
        this.hide()
    }

    private fun onCancelClick(event: ClickEvent) {
        this.gameToLoad = ""
        this.exitState = LoadGameExitStates.CANCEL
        this.hide()
    }

    private fun onDeleteClick(event: ClickEvent) {
        this.confirmDeleteDialog = ConfirmDeleteSaveDialog(this.gameList!!.getSelected().toString(), this.skinInternal)
        this.confirmDeleteDialog!!.addDialogListener({ args -> this.onConfirmDeleteSaveDialogClose(args) }, ObservableDialog.DialogEvents.HIDDEN)
        this.confirmDeleteDialog!!.show(this.getStage())
        this.confirmDeleteDialog!!.setSize(this.widthInternal, this.heightInternal)
    }

    private fun onConfirmDeleteSaveDialogClose(args: ObservableDialog.DialogEventArgs) {
        if (this.confirmDeleteDialog!!.exitState === ConfirmDeleteSaveDialog.ConfirmDeleteExitStates.DELETE) {
            EuropaGame.game.deleteSave(this.gameList!!.getSelected().toString())
            this.gameList!!.setItems(GdxArray(EuropaGame.game.getSaveNames()))
            this.okButton!!.setDisabled(EuropaGame.game.getSaveNames().size() == 0)
        }
    }


    // Nested Classes
    private class ConfirmDeleteSaveDialog(public val saveToDelete: String, private val skinInternal: Skin) : ObservableDialog(LoadGameDialog.ConfirmDeleteSaveDialog.DIALOG_NAME, skinInternal.get(MainMenuScreen.POPUP_DIALOG_STYLE_KEY, javaClass<Window.WindowStyle>())) {

        // Enumerations
        public enum class ConfirmDeleteExitStates {

            DELETE,
            CANCEL
        }

        private var titleLabelInternal: Label? = null
        private var mainTable: Table? = null
        private var yesButton: TextButton? = null
        private var noButton: TextButton? = null

        public var exitState: ConfirmDeleteExitStates? = null
            private set


        init {

            this.initComponents()
        }


        // Public Methods
        override fun setSize(width: Float, height: Float) {
            super.setSize(width, height)
        }


        // Private Methods
        private fun initComponents() {
            this.titleLabelInternal = Label("Are you sure you want to delete the save game \"" + this.saveToDelete + "\"?", skinInternal.get(MainMenuScreen.INFO_STYLE_KEY, javaClass<Label.LabelStyle>()))
            this.titleLabelInternal!!.setWrap(true)
            this.titleLabelInternal!!.setAlignment(Align.center)

            this.yesButton = TextButton("Yes", skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<TextButton.TextButtonStyle>()))
            this.yesButton!!.addListener(object : ClickListener() {
                override fun clicked(event: InputEvent?, x: Float, y: Float) {
                    if (event!!.getButton() == Input.Buttons.LEFT && !this@ConfirmDeleteSaveDialog.yesButton!!.isDisabled()) {
                        this@ConfirmDeleteSaveDialog.onYesButtonClick()
                    }
                    this@ConfirmDeleteSaveDialog.yesButton!!.setChecked(false)
                }
            })

            this.noButton = TextButton("No", skinInternal.get(MainMenuScreen.DEFAULT_KEY, javaClass<TextButton.TextButtonStyle>()))
            this.noButton!!.addListener(object : ClickListener() {
                override fun clicked(event: InputEvent?, x: Float, y: Float) {
                    if (event!!.getButton() == Input.Buttons.LEFT && !this@ConfirmDeleteSaveDialog.noButton!!.isDisabled()) {
                        this@ConfirmDeleteSaveDialog.onNoButtonClick()
                    }
                    this@ConfirmDeleteSaveDialog.noButton!!.setChecked(false)
                }
            })

            this.mainTable = Table()
            this.mainTable!!.add<Label>(this.titleLabelInternal).colspan(2).width(MainMenuScreen.DIALOG_WIDTH.toFloat()).expandX().fillX()
            this.mainTable!!.row()
            this.mainTable!!.add<TextButton>(this.noButton).width(MainMenuScreen.DIALOG_BUTTON_WIDTH.toFloat()).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).right()
            this.mainTable!!.add<TextButton>(this.yesButton).width(MainMenuScreen.DIALOG_BUTTON_WIDTH.toFloat()).space(MainMenuScreen.COMPONENT_SPACING.toFloat()).left()

            this.mainTable!!.padLeft(MainMenuScreen.TABLE_PADDING.toFloat())
            this.mainTable!!.padRight(MainMenuScreen.TABLE_PADDING.toFloat())

            this.mainTable!!.background(this.skinInternal.get(MainMenuScreen.POPUP_BACKGROUND_KEY, javaClass<SpriteDrawable>()))
            this.getContentTable().add<Table>(this.mainTable).row()
        }

        private fun onYesButtonClick() {
            this.exitState = ConfirmDeleteExitStates.DELETE
            this.hide()
        }

        private fun onNoButtonClick() {
            this.exitState = ConfirmDeleteExitStates.CANCEL
            this.hide()
        }

        companion object {


            // Constants
            private val DIALOG_NAME = ""
        }

    }// Properties

    companion object {


        // Constants
        private val DIALOG_NAME = ""


        // Static Methods
        private fun getDefaultSkin(): Skin {
            return MainMenuScreen.createMainMenuSkin()
        }
    }

}
